partiql-lang/src/main/kotlin/org/partiql/lang/graph/ExternalGraphReader.kt

```diff
@@ -34,4 +46,98 @@ object ExternalGraphReader {
         val graphStr = graphFile.readText()
         validate(graphStr)
     }
+
+    /** Given an IonValue, validates that it is a graph in accordance with graph.isl schema
+     *  and loads it into memory as a SimpleGraph. */
+    internal fun read(graphIon: IonValue): SimpleGraph {
+        validate(graphIon)
+        return readGraph(graphIon)
+    }
+    internal fun read(graphStr: String): SimpleGraph {
+        val graphIon = ion.singleValue(graphStr)
+        validate(graphIon)
+        return readGraph(graphIon)
+    }
+
+    /* Entry point into the implementation of graph reading.
+     * It is assumed that the input Ion value has been validated w.r.t. graph ISL,
+     * so that defensive error checks are not needed.
+     */
+    internal fun readGraph(graphIon: IonValue): SimpleGraph {
+        val g = graphIon as IonStruct
+        val nds = g.get("nodes") as IonList
+        val eds = g.get("edges") as IonList
+        val nodes = readNodes(nds)
+        val (directed, undirected) = EdgeReader(nodes).readEdges(eds)
+        return SimpleGraph(nodes.values.toList(), directed, undirected)
+    }
+
+    /* Returns a map from node IDs in the source graph to newly-built in-memory nodes.
+    *  This map is used later to recognize node IDs used in edge definitions.
+    */
+    internal fun readNodes(nodesList: IonList): Map<String, SimpleGraph.Node> {
+        val pairs = nodesList.toList().map { readNode(it as IonStruct) }
+        val duplicates = pairs.map { it.first }.groupBy { it }.filterValues { it.size > 1 }.keys
+        if (duplicates.isNotEmpty()) throw GraphReadException(
+            "Identifiers used for more than one node: ${duplicates.joinToString()}."
```

Comment:
Perhaps include the duplicate identifier names in the error message

Reply:
`${duplicates.joinToString()}` does just that?

Reply:
Scheduling an optometrist appointment now... 😆